### PR TITLE
Fix installation directory.

### LINKED ISSUE #6625: Installing to non-default directory creates incorrect references in lib files
When using a non-default value for CMAKE_INSTALL_PREFIX in CMake, OpenCV lib files have incorrect directory references which makes it impossible for the libraries to load required libraries. Attempting to run the executable results in the following error:

> dyld: Library not loaded: libopencv_highgui.3.1.dylib
>   Referenced from: /Users/steve/Library/Developer/Xcode/DerivedData/deskewer-eqsavpdgycukoedxisschlsfozor/Build/Products/Debug/deskewer
>   Reason: image not found
### Please state the information for your system
- OpenCV version: 3.x
- Host OS: Mac OS X 10.11.5
- Compiler & CMake: clang-703.0.29 & CMake 3.1.0
### Library references in the file libopencv_video.3.1.0.dylib:

```
$ otool -L libopencv_video.3.1.0.dylib
libopencv_video.3.1.0.dylib:
    libopencv_video.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
    libopencv_imgproc.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
    libopencv_core.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
    /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
    /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

Note that the referenced dylib files have no path.

I've found a solution and will submit a pull request later today.

### PR DESCRIPTION
**Pull request by SSteve**
resolves #6625

### What does this PR change?
Set CMAKE_INSTALL_NAME_DIR when CMAKE_INSTALL_PREFIX isn't default value of "/usr/local" in a Unix/macOS/CygWin installation.

CMAKE_INSTALL_NAME_DIR  causes the non-default installation directory to be added to the library file references. If this isn't done, it causes one of two problems:

1. If OpenCV libraries are already installed in /usr/local, these will be loaded instead of the versions in the non-default location or
2. If OpenCV libraries aren't already installed in /usr/local, the program will not run because the libraries in the non-default location won't be found.

###Results of otool before the change:
```
$ otool -L libopencv_video.3.1.0.dylib
libopencv_video.3.1.0.dylib:
	libopencv_video.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	libopencv_imgproc.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	libopencv_core.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

###Results of otool after the change:
```
$ otool -L libopencv_video.3.1.0.dylib
libopencv_video.3.1.0.dylib:
	/usr/local/stevecv/lib/libopencv_video.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	/usr/local/stevecv/lib/libopencv_imgproc.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	/usr/local/stevecv/lib/libopencv_core.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```

### CODE REVIEW

**Comment by mshabunin** [comment]
Did you try to build using cmake option `-DCMAKE_MACOSX_PATH=ON`?
```
$ otool -L lib/libopencv_video.dylib                                                                 
lib/libopencv_video.dylib:
        @rpath/libopencv_video.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
        @rpath/libopencv_imgproc.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
        @rpath/libopencv_core.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)

$ otool -L install/lib/libopencv_video.dylib                                                         
install/lib/libopencv_video.dylib:
        @rpath/libopencv_video.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
        @rpath/libopencv_imgproc.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
        @rpath/libopencv_core.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
        /usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 120.1.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1226.10.1)
```
But I'm not sure how it should work: [Apple documentation](https://developer.apple.com/library/mac/documentation/DeveloperTools/Conceptual/DynamicLibraries/100-Articles/RunpathDependentLibraries.html), [cmake documentation](https://cmake.org/cmake/help/v3.0/prop_tgt/MACOSX_RPATH.html)

**Comment by SSteve** [comment]
@mshabunin I'll give that a try and report back but I won't be able to do it until next week.

**Comment by SSteve** [comment]
@mshabunin I tried `CMAKE_MACOSX_RPATH` (it's not `CMAKE_MACOSX_PATH`). It did prepend `@rpath/` to the library path but I still get "dyld: Library not loaded" when attempting to load the libraries in `/usr/local/stevecv`.

```
$ otool -L libopencv_highgui.dylib
libopencv_highgui.dylib:
	@rpath/libopencv_highgui.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	@rpath/libopencv_videoio.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	/usr/local/opt/ffmpeg/lib/libavcodec.56.dylib (compatibility version 56.0.0, current version 56.60.100)
...
	@rpath/libopencv_imgcodecs.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	@rpath/libopencv_imgproc.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
	@rpath/libopencv_core.3.1.dylib (compatibility version 3.1.0, current version 3.1.0)
```

**Comment by alalek** [comment]
Could you try [this](https://cmake.org/cmake/help/v3.5/variable/CMAKE_INSTALL_NAME_DIR.html):
```
cmake -DCMAKE_INSTALL_NAME_DIR=/usr/local/stevecv/lib ...
```

**Comment by SSteve** [comment]
Setting `CMAKE_INSTALL_NAME_DIR` solves the problem. I modified my change to OpenCVModule.cmake so that it automatically sets `CMAKE_INSTALL_NAME_DIR` to `${CMAKE_INSTALL_PREFIX}/lib` when `CMAKE_INSTALL_PREFIX` isn't set to "/usr/local".

File: cmake/OpenCVModule.cmake [hunk 1]
```diff
@@ -75,6 +75,12 @@ set(OPENCV_MODULES_DISABLED_AUTO  "" CACHE INTERNAL "List of OpenCV modules impl
 set(OPENCV_MODULES_DISABLED_FORCE "" CACHE INTERNAL "List of OpenCV modules which can not be build in current configuration")
 unset(OPENCV_WORLD_MODULES CACHE)
 
+# set CMAKE_INSTALL_NAME_DIR if CMAKE_INSTALL_PREFIX isn't default value of "/usr/local"
+if(UNIX AND NOT ${CMAKE_INSTALL_PREFIX} STREQUAL "/usr/local")
```
Comment (alalek) [inline cmake/OpenCVModule.cmake]: Does it work with `AND NOT DEFINED CMAKE_INSTALL_NAME_DIR` additional condition (we should not forbid to configure settings via command-line)?

**Comment by vpisarev** [comment]
this PR will be reopened because of CI issues

**Pull request closed without merging**